fish-rust/src/builtins/tests/string_tests.rs

```diff
@@ -0,0 +1,305 @@
+        [$($x:tt),* $(,)?] => { [$(test_cases!$x),*] };
+    }
+
+    // TODO: these should be individual tests, not all in one, port when we can run these with `cargo test`
+    fn string_test(mut args: Vec<&wstr>, expected_rc: Option<i32>, expected_out: &wstr) {
+        let parser: &mut parser_t = unsafe { &mut *parser_t::principal_parser_ffi() };
+        let mut streams = ffi::make_test_io_streams_ffi();
+        let mut io = crate::builtins::shared::io_streams_t::new(streams.pin_mut());
+
+        let rc = string(parser, &mut io, args.as_mut_slice()).expect("string failed");
+
+        assert_eq!(expected_rc.unwrap(), rc, "string builtin returned unexpected return code");
+
+        let string_stream_contents = &ffi::get_test_output_ffi(&streams);
+        let actual = escape_string(&string_stream_contents.from_ffi(), EscapeStringStyle::default());
+        let expected = escape_string(expected_out, EscapeStringStyle::default());
+        assert_eq!(expected, actual, "string builtin returned unexpected output");
+    }
+
+    let tests = test_cases![
+        (["string", "escape"], STATUS_CMD_ERROR, ""),
+        (["string", "escape", ""], STATUS_CMD_OK, "''\n"),
+        (["string", "escape", "-n", ""], STATUS_CMD_OK, "\n"),
+        (["string", "escape", "a"], STATUS_CMD_OK, "a\n"),
+        (["string", "escape", "\x07"], STATUS_CMD_OK, "\\cg\n"),
+        (["string", "escape", "\"x\""], STATUS_CMD_OK, "'\"x\"'\n"),
+        (["string", "escape", "hello world"], STATUS_CMD_OK, "'hello world'\n"),
+        (["string", "escape", "-n", "hello world"], STATUS_CMD_OK, "hello\\ world\n"),
+        (["string", "escape", "hello", "world"], STATUS_CMD_OK, "hello\nworld\n"),
+        (["string", "escape", "-n", "~"], STATUS_CMD_OK, "\\~\n"),
+
+        (["string", "join"], STATUS_INVALID_ARGS, ""),
+        (["string", "join", ""], STATUS_CMD_ERROR, ""),
+        (["string", "join", "", "", "", ""], STATUS_CMD_OK, "\n"),
+        (["string", "join", "", "a", "b", "c"], STATUS_CMD_OK, "abc\n"),
+        (["string", "join", ".", "fishshell", "com"], STATUS_CMD_OK, "fishshell.com\n"),
+        (["string", "join", "/", "usr"], STATUS_CMD_ERROR, "usr\n"),
+        (["string", "join", "/", "usr", "local", "bin"], STATUS_CMD_OK, "usr/local/bin\n"),
+        (["string", "join", "...", "3", "2", "1"], STATUS_CMD_OK, "3...2...1\n"),
+        (["string", "join", "-q"], STATUS_INVALID_ARGS, ""),
+        (["string", "join", "-q", "."], STATUS_CMD_ERROR, ""),
+        (["string", "join", "-q", ".", "."], STATUS_CMD_ERROR, ""),
+
+        (["string", "length"], STATUS_CMD_ERROR, ""),
+        (["string", "length", ""], STATUS_CMD_ERROR, "0\n"),
+        (["string", "length", "", "", ""], STATUS_CMD_ERROR, "0\n0\n0\n"),
+        (["string", "length", "a"], STATUS_CMD_OK, "1\n"),
+
+    // #if WCHAR_T_BITS > 16
+        (["string", "length", "\u{2008A}"], STATUS_CMD_OK, "1\n"),
+    // #endif
```

Comment:
> This #if we can drop because we have stronger guarantees now.
In Rust, we only deal with platform-specific wchar_t when converting to and from OS strings using mbrtowc which is not what we test here.

```suggestion
        (["string", "length", "\u{2008A}"], STATUS_CMD_OK, "1\n"),
```